### ISSUE TITLE
Survey existing FTP servers and clients to figure out how LIST works

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7141]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1333|
|Type|enhancement|
|Created|2005-11-14 01:29:16Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-list-flags-1333-2|

Attachments:

* [1333-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/aee/aee6e7cba08370f8e346c2a6268372c40be71fb7/0f5041c885c7eeb128bebe1fb4874f246fda6408.diff) (2074 bytes) - added by adiroiban on 2013-03-05 19:33:06Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1333 1333
type__enhancement enhancement
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__branches_ftp_list_flags_1333_2 branches/ftp-list-flags-1333-2
branch_author__spiv__adiroiban spiv, adiroiban
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1131931756000000 1131931756000000
changetime__1362593808000000 1362593808000000
version__ 
owner__ 
cc__exarkun cc__spiv cc__itamarst cc__jknight cc__jesstess cc__adiroiban
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Currently we special case a few options in ftp_LIST.  This sucks and we should
do something better.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

Beside the current code:

```
        # bug in konqueror
        if path == "-a":
            path = ''
        # bug in gFTP 2.0.15
        if path == "-aL":
            path = ''
        # bug in Nautilus 2.10.0
        if path == "-L":
            path = ''
        # bug in ange-ftp
        if path == "-la":
            path = ''
```

I have also discovered the following clients:

```
        # bug in FireFTP
        if path == "-al":
            path = ''

        # bug in BareFTP
        if path == "-La":
            path = ''
```

Maybe we can use a regex or something simple like:

```
if path.lower() in ['-a', '-l', '-la', '-al']:
    path = ''
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|spiv set owner to spiv<br>spiv set status to `assigned`|
|-|-|

Replying to [adiroiban](#note_4):
> Maybe we can use a regex or something simple like:
> 
> ```
> if path.lower() in ['-a', '-l', '-la', '-al']:
>     path = ''
> ```

I like this approach.  Obviously it needs tests etc.  I'll make a patch, if I don't fall asleep or otherwise get distracted first :)

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|spiv commented|
|-|-|

(In [34787]) Branch to ftp-list-flags-1333

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv removed owner<br>spiv set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to spiv|
|-|-|

1. Please name the test test_LISTWithBinLsFlags and add a docstring.

2. This needs a NEWS fragment.

Thanks, merge once those points are fixed!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

[#5791](https://github.com/twisted/twisted/issues/5791) has been closed a duplicate of this bug.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban removed owner|
|-|-|

I have attached a patch on top of branch 1333

I have renamed the test and add docstring. Also added docstring and renaming for neighbor tests.

It looks like the branch already has a news file.

----

Please switch to git/github. Working with SVN is so slow, Trac code navigation is slow and uploading patched to tickets is not fun.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_12"></a>|@tomprince commented|
|-|-|

(In [37410]) Branching to ftp-list-flags-1333-2.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince set status to `closed`|
|-|-|

(In [37413]) Merge ftp-list-flags-1333-2: Handle bogus options to ftp's LIST command.

Author: spiv, adiroiban
Reviewers: therve, tom.prince
Fixes: [#7141](https://github.com/twisted/twisted/issues/7141)

The twisted.protocols.ftp.FTP server now treats "LIST -La", "LIST -al", and all
other combinations of ordering and case of the "-l" and "-a" flags the same: by
ignoring them rather than treating them as a pathname.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_14"></a>|@tomprince commented|
|-|-|

(I accidentally put my review on [#5411](https://github.com/twisted/twisted/issues/5411) by mistake:)

Minor change: Prefer assertIn to assertTrue(... in ...).

I've fixed this up and merged forward.

Will merge once  build results are in.
